src/mgr.sol

```diff
@@ -186,128 +205,131 @@ contract TinlakeManager {
-        vat.frob(ilk, address(this), address(this), address(this), -int256(wad), 0);
-        vat.slip(ilk, address(this), -int256(wad));
         gem.transfer(msg.sender, wad);
         emit Exit(wad);
     }
 
     // draw & wipe call daiJoin.exit/join immediately
-    function draw(uint256 wad) public operatorOnly {
+    function draw(uint256 wad) public auth {
         require(safe && live, "TinlakeManager/bad-state");
-        (, uint256 rate, , , ) = vat.ilks(ilk);
-        uint256 dart = divup(mul(RAY, wad), rate);
-        require(int256(dart) >= 0, "TinlakeManager/overflow");
-        vat.frob(ilk, address(this), address(this), address(this), 0, int256(dart));
-        daiJoin.exit(msg.sender, wad);
+        urn.draw(wad);
+        dai.transferFrom(address(this), msg.sender, wad);
         emit Draw(wad);
     }
 
-    function wipe(uint256 wad) public operatorOnly {
+    function wipe(uint256 wad) public {
         require(safe && live, "TinlakeManager/bad-state");
-        dai.transferFrom(msg.sender, address(this), wad);
-        daiJoin.join(address(this), wad);
-        (,uint256 rate, , , ) = vat.ilks(ilk);
-        uint256 dart = mul(RAY, wad) / rate;
-        require(dart <= 2 ** 255, "TinlakeManager/overflow");
-        vat.frob(ilk, address(this), address(this), address(this), 0, -int256(dart));
+        dai.transferFrom(msg.sender, address(urn), wad);
+        urn.wipe(wad);
         emit Wipe(wad);
     }
 
-    // --- Administration ---
-    function setOperator(address newOperator) external operatorOnly  {
-        operator = newOperator;
-        emit SetOperator(newOperator);
+    function free(uint256 wad) public auth {
+        urn.quit();
     }
 
-    function migrate(address dst) public auth  {
-        vat.hope(dst);
+    // --- Administration ---
+    function migrate(address dst) public {
+        require(can[msg.sender] == 1 || wards[msg.sender] == 1, "TinlakeMgr/auth-or-gov-only");
+
         dai.approve(dst, uint256(-1));
         gem.approve(dst, uint256(-1));
         live = false;
         emit Migrate(dst);
     }
 
+    function file(bytes32 what, bytes32 data) public auth {
+        emit File(what, data);
+        if (what == "ilk") {
+          ilk = what;
+        } else revert("TinlakeMgr/file-unknown-param");
+    }
+
     function file(bytes32 what, address data) public auth {
-        require(live, "TinlakeManager/not-live");
         emit File(what, data);
-        if (what == "vow") vow = data;
-        else if (what == "daiJoin") daiJoin = JoinLike(data);
-        else if (what == "end")  end = EndLike(data);
-        else revert("Vat/file-unrecognized-param");
+        if (what == "urn") {
+            urn = MIP21UrnLike(data);
+            dai.approve(data, uint256(-1));
+        }
+        else if (what == "owner") {
+            owner = data;
+        }
+        else if (what == "rwaToken") {
+            rwaToken = GemLike(data);
+        }
```

Comment:
This is also immutable.  Also, you can get this from the `urn.gemJoin().gem() `.
```suggestion
```

Reply:
Nice!